src/common/shared_cache.hpp

```diff
@@ -348,14 +348,26 @@ class SharedLRU {
     VPtr val;
     list<VPtr> to_release;
     {
-      std::lock_guard l{lock};
-      typename map<K, pair<WeakVPtr, V*>, C>::iterator actual =
-	weak_refs.lower_bound(key);
-      if (actual != weak_refs.end() && actual->first == key) {
-        if (existed) 
-          *existed = true;
+      typename map<K, pair<WeakVPtr, V*>, C>::iterator actual;
+      std::unique_lock l{lock};
+      cond.wait(l, [this, &key, &actual, &val] {
+	  actual = weak_refs.lower_bound(key);
+	  if (actual != weak_refs.end() && actual->first == key) {
+	    if (val = actual->second.first.lock(); val) {
```

Comment:
@gregsfortytwo this is a C++17 feature. namely "if statement with Initializer", see https://en.cppreference.com/w/cpp/language/if#If_Statements_with_Initializer .

and IMO, it does help with clarity. and the code is more concise this way.

Reply:
Mmm, I'm pretty sure this is not the "if statement with initializer"!
We have been able to do assignments within an if statement for a very long time, because assignments "return" the object/primitive after it has been assigned. That's why I'm saying we could just do ``` if (val = actual->second.first.lock()```

I'm not sure if we've been able to combine multiple statements within an if-evaluation, but I think we have? Maybe that's new — the big part I'm seeing out of that C++17-ism is that you can declare new variables which are scoped inside the if/else block not outside of it. :)

Reply:
@gregsfortytwo yes, we can. and it's standard compliant. and they are equivalent. just that, at the first glance, people might think we are making a mistake like
```c++
if (today = Friday) {
  // happy friday!
}
```
but we are not.

also i want to elaborate a little bit why i think it is "if statement with initializer".
```c++
if (val = actual->second.first.lock(); val) {
```
is indeed following a valid and well defined syntax in C++17.

> I'm not sure if we've been able to combine multiple statements within an if-evaluation, but I think we have? 

that has been a GCC/Clang extension to the C standard and C++98 standard for a long time. see https://gcc.gnu.org/onlinedocs/gcc/Statement-Exprs.html#Statement-Exprs

> A compound statement enclosed in parentheses may appear as an expression in GNU C

which enables us to put more stuff into a `()`. in this case, they are
```c++
val = actual->second.first.lock(); val
```
it is not a valid expression in pre-C++17 era. and C++17 defines the `if` statement as

> if constexpr{opt} ( init-statement_{opt} condition ) statement

where "init-statement" can also be an expression as well as a simple declaration as you noticed. in this case, the expression is just an assignment expression. that's why i think `val = actual->second.first.lock(); val` is valid.

> the big part I'm seeing out of that C++17-ism is that you can declare new variables which are scoped inside the if/else block not outside of it. 

agreed. and IMO, it also standarize/formalizes a typical use case of GCC extension to C language.